Implement Status for Application CRD

barney-s opened the pull request:
Implemented a draft version of Status for application crd.

Reference:
https://github.com/kubernetes-sigs/application/pull/77
https://github.com/kubernetes-sigs/application/pull/79 (closest to this)
https://github.com/kubernetes-sigs/application/issues/78

1. Manually vendor kubesdk
2. Use generic reconciler in Application controller loop.
3. Hack Mutate to return observed as expected (need a slightly better way)
4. Use meta.Status

Known Issues:
1. Need to ReSync with open proposals once they mature.
2. Periodic syncs(2m) to update status. Need to add dynamic watchers.
3. Unknown resources need to be handled.


Example Status with these changes:
```yaml
status:
    components:
    - link: /apis/apps/v1beta2/namespaces/default/statefulsets/mc-base-nfs
      name: mc-base-nfs
      status: Ready
    - link: /apis/apps/v1beta2/namespaces/default/statefulsets/mc-base-mysql
      name: mc-base-mysql
      status: Ready
    - link: /apis/apps/v1beta1/namespaces/default/statefulsets/mc-base-mysql
      name: mc-base-mysql
      status: Ready
    - link: /apis/apps/v1beta1/namespaces/default/statefulsets/mc-base-nfs
      name: mc-base-nfs
      status: Ready
    - link: /apis/apps/v1/namespaces/default/statefulsets/mc-base-mysql
      name: mc-base-mysql
      status: Ready
      sts:
        currentcount: 1
        progress: 0
        readycount: 1
        replicas: 1
    - link: /apis/apps/v1/namespaces/default/statefulsets/mc-base-nfs
      name: mc-base-nfs
      status: Ready
      sts:
        currentcount: 1
        progress: 0
        readycount: 1
        replicas: 1
    - link: /api/v1/namespaces/default/services/mc-base-sql
      name: mc-base-sql
      status: Ready
    - link: /api/v1/namespaces/default/services/mc-base-nfs
      name: mc-base-nfs
      status: Ready
    - link: /apis/policy/v1beta1/namespaces/default/poddisruptionbudgets/mc-base-mysql
      name: mc-base-mysql
      pdb:
        currenthealthy: 1
        desiredhealthy: 1
      status: Ready
    - link: /apis/policy/v1beta1/namespaces/default/poddisruptionbudgets/mc-base-nfs
      name: mc-base-nfs
      pdb:
        currenthealthy: 1
        desiredhealthy: 1
      status: Ready
    - link: /api/v1/namespaces/default/secrets/mc-base-sql
      name: mc-base-sql
      status: Ready
    conditions:
    - lastTransitionTime: 2018-12-07T22:53:23Z
      lastUpdateTime: 2018-12-07T22:53:23Z
      message: all components ready
      reason: ComponentsReady
      status: "True"
      type: Ready
    - lastTransitionTime: 2018-12-07T22:46:00Z
      lastUpdateTime: 2018-12-07T22:46:00Z
      message: No error seen
      reason: NoError
      status: "False"
      type: Error
```

barney-s commented:
@kow3ns , @konryd , @mortent , @huyhg - Please review

barney-s commented:
Willing to iterate the code once the proposals are agreed upon.

mortent commented on pkg/apis/app/v1beta1/application.go:
```diff
@@ -0,0 +1,164 @@
+	"sigs.k8s.io/kubesdk/pkg/status"
+	"strings"
+)
+
```
Some better comments on the functions in this file would be useful. It is unclear to me how each of them fit into the reconcile loop.

barney-s commented on Makefile:
```diff
@@ -16,6 +16,10 @@ manager: generate fmt vet
 
+# Run against the configured Kubernetes cluster in ~/.kube/config
+debug: generate fmt vet
+	dlv debug cmd/manager/main.go
```
debugger

barney-s commented on pkg/apis/app/v1beta1/application_types.go:
```diff
@@ -132,8 +131,10 @@ type InfoItem struct {
 	ValueFrom *InfoItemSource `json:"valueFrom,omitempty"`
 }
 
+// InfoItemType is a string
```
done

barney-s commented on Makefile:
```diff
@@ -16,6 +16,10 @@ manager: generate fmt vet
 run: generate fmt vet
 	go run ./cmd/manager/main.go
 
+# Run against the configured Kubernetes cluster in ~/.kube/config
```
changed it to debug

EronWright commented on pkg/apis/app/v1beta1/application.go:
```diff
@@ -0,0 +1,169 @@
+	return nil
+}
+
+//DependantResources - returns dependent rsrc
```
Do you mean 'dependent', 'dependant', or both?  The latter is a legit British word.

barney-s replied:
will change to "dependent". Had similar comment earlier as well. Guess i am still used to the British spelling.

barney-s replied:
done

kow3ns commented:
/approve

ant31 commented:
Is it in the state to be merged @barney-s (as the title indicates 'draft') ?

prydonius commented on config/crds/app_v1beta1_application.yaml:
```diff
@@ -189,6 +189,75 @@ spec:
+                    type: string
+                  name:
+                    type: string
+                  pdb:
```
Should this be the short name? I think using the full name would make more sense.

barney-s replied:
can change it.

barney-s replied:
done

barney-s commented:
> Is it in the state to be merged @barney-s (as the title indicates 'draft') ?

Will update the CL to reflect the mergeable status by this week.
Hold merge till then.

barney-s commented:
> > Is it in the state to be merged @barney-s (as the title indicates 'draft') ?
> 
> Will update the CL to reflect the mergeable status by this week.
> Hold merge till then.

Ready for merge.
There are a few Follow up which can be resolved as issues
1. I kept this comment open. Not sure if we need to embed conditions of tracked in app status.
   https://github.com/kubernetes-sigs/application/pull/81#pullrequestreview-193802579
2. shows 2 entries for statefulsets (v1beta2, v1) in status array
3. reconcile loop sometimes resyncs multiple times when single time should be sufficient.

barney-s commented on pkg/genericreconciler/genericreconciler.go:
```diff
@@ -0,0 +1,351 @@
+	for _, obs := range observables {
+		var resources []resource.Object
+		if obs.Labels != nil {
+			//log.Printf("   >>>list: %s labels:[%v]", reflect.TypeOf(obs.ObjList).String(), obs.Labels)
```
ack

barney-s commented on pkg/genericreconciler/genericreconciler.go:
```diff
@@ -0,0 +1,351 @@
+		eKind := reflect.TypeOf(e.Obj).String()
+		eRsrcInfo := eNamespace + "/" + eKind + "/" + eName
+		for _, o := range observed.Items() {
+			if (eName == o.Obj.GetName()) &&
```
ack

barney-s commented on pkg/genericreconciler/v1alpha1/crfoo.go:
```diff
@@ -0,0 +1,221 @@
+/*
```
removed.

barney-s commented on pkg/status/types.go:
```diff
@@ -0,0 +1,141 @@
+
+// Statefulset is a generic status holder for stateful-set
+// +k8s:deepcopy-gen=true
+type Statefulset struct {
```
These are extended information for core types. Envisioning extending this as we gain more feedback.

barney-s replied:
removing for now. to keep it simple.

barney-s commented on cmd/manager/main.go:
```diff
@@ -41,8 +42,9 @@ func main() {
 	}
 
 	// Create a new Cmd to provide shared dependencies and start components
+	syncperiod := time.Minute * 2
```
just a meaningful default. Alternative is to setup dynamic watchers for all types in all application objects which may quickly escalate to all possible API objects !!

kow3ns replied:
Should probably put that in a comment to make the reasoning more obvious.

barney-s replied:
done

barney-s commented on pkg/apis/app/v1beta1/application.go:
```diff
@@ -0,0 +1,169 @@
+
+// Differs returns true if the resource needs to be updated
+func (a *Application) Differs(expected metav1.Object, observed metav1.Object) bool {
+	return true
```
this is the semantic comparison. true is a lazy implementation punting the update filtering to api server. How the api server does this is something i intend to understand and implement here.

barney-s commented on pkg/genericreconciler/genericreconciler.go:
```diff
@@ -0,0 +1,351 @@
+	// Not all k8s objects have Spec
+	// example ConfigMap
+	// TODO strategic merge patch diff in generic controller loop
+	e := reflect.Indirect(reflect.ValueOf(o1)).FieldByName("Spec")
```
this is a gate for sematic difference check. if this doesnt pass we call semantic equivalence check via Differs() interface.

barney-s commented on pkg/genericreconciler/genericreconciler.go:
```diff
@@ -0,0 +1,351 @@
+	var expected, observed, dependent *resource.ObjectBag
+
+	// Get dependenta objects
+	stage := "dependent resources"
```
done.

barney-s commented on pkg/genericreconciler/genericreconciler.go:
```diff
@@ -0,0 +1,351 @@
+	stage := "dependent resources"
+	dependent, err = gr.observe(resource.ObservablesFromObjects(gr.Scheme, c.DependentResources(c.CR), c.Labels())...)
+
+	if err == nil && dependent != nil {
```
done.

barney-s commented:
> It would be useful if we had one proposal with consensus on how we want status for Application CRD to work. Based on the three proposals linked here, it is unclear to me how we want this to work.

This is based on #79

barney-s commented on pkg/status/types.go:
```diff
@@ -0,0 +1,141 @@
+
+// ExtendedStatus is a holder of additional status for well known types
+// +k8s:deepcopy-gen=true
+type ExtendedStatus struct {
```
done.

barney-s commented on pkg/apis/app/v1beta1/application_types.go:
```diff
@@ -78,9 +79,7 @@ type ApplicationSpec struct {
-	// ObservedGeneration is used by the Application Controller to report the last Generation of an Application
-	// that it has observed.
-	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	status.Meta `json:",inline"`
```
status/condition.go methods and UpdateStatus methods are Meta struct methods. 
Extracting UpdateStatus is trivial. But moving all condition functions would need non-trivial refactoring.

Keeping it as a separate package (status) to provide an example for other controllers.

barney-s replied:
merged.

barney-s commented on pkg/status/types.go:
```diff
@@ -0,0 +1,141 @@
+/*
```
Would renaming Meta to something less confusing help.
Inlining is an artifact of having status code reusable across controllers. (sdk)

kow3ns replied:
In this code base there is only one implementation. I don't anticipate anyone vendoring directly from this code base. I don't see why the refactoring would be particularly difficult. If an application sdk is offered as an open source package in the future we should rebase on top of it. In the mean time, I'd really prefer to see all of the relevant types pertaining to application in a single place with a uniform implementation.

barney-s replied:
done moving the status bits into application api

barney-s commented:
> See the with respect to implementing status in its own package and importing it. I think the complexity along with the mirroring of the meta types in api machinery might be confusing. Also its not generally the way that k8s apis are implements.
> You need to fix your code headers.
> So of the code could do with refactoring to a more golang idiomatic happy path style.
> Please remove all commented and uncommented printf debugging statements and all commented out dead code.

- removed dead code
- fixed authors
- refactored for more happy path and less nesting
- removed extended status for known types

with respect to status inline, would renaming meta to something less conflicting help ? Inlining is an artifact of having status package reusable across controllers (sdk)

barney-s commented:
@kow3ns 
Moved status bits into application api directly.
Please review and approve.

barney-s commented:
> Shouldn't the ComponentList contain the group, version, kind and namespace or is the idea that this can be extracted from the link?

A good suggestion. Can do this as a follow up issue.

barney-s commented:
@huyhg  @mortent @ant31 @prydonius 
Could one of you please review and LGTM ?

huyhg commented on pkg/apis/app/v1beta1/application_types.go:
```diff
@@ -202,8 +278,10 @@ type IngressSelector struct {
 	Path string `json:"path,omitempty"`
 }
 
+// ApplicationAssemblyPhase tracks the Application CRD phases: pending, succeded, failed
```
Do we take this into account?

huyhg commented on pkg/apis/app/v1beta1/status.go:
```diff
@@ -0,0 +1,91 @@
+
+// Resource specific logic -----------------------------------
+
+// Statefulset
```
Just to confirm: more will be added, once the proposal PR is approved, right?

kow3ns commented:
@huyhg can you lgtm please I already approved

huyhg commented:
@kow3ns Already LGTM'ed.

kow3ns commented:
you need to actually "/lgtm" so the label applies and it merges and it has to happen after any commits.

huyhg commented:
/lgtm

prydonius commented:
/lgtm

cmoulliard commented:
Is this PR documented ? Without doc, nobody will be able to use the approach proposed here to handle status ?
